ambry-replication/src/main/java/com/github/ambry/replication/ReplicationEngine.java

```diff
@@ -311,25 +311,72 @@ protected void removeRemoteReplicaInfoFromReplicaThread(List<RemoteReplicaInfo>
    * a {@link ReplicaThread} will be selected by {@link ReplicationEngine#getReplicaThreadIndexToUse(String)}.
    * Create threads pool for a DC if not exists.
    * @param remoteReplicaInfos List of {@link RemoteReplicaInfo} to add.
-   * @param startThread if threads need to be started when create.
+   * @param startThread if threads need to be started when created.
    */
   protected void addRemoteReplicaInfoToReplicaThread(List<RemoteReplicaInfo> remoteReplicaInfos, boolean startThread) {
     for (RemoteReplicaInfo remoteReplicaInfo : remoteReplicaInfos) {
-      DataNodeId dataNodeIdToReplicate = remoteReplicaInfo.getReplicaId().getDataNodeId();
-      String datacenter = dataNodeIdToReplicate.getDatacenterName();
-      List<ReplicaThread> replicaThreads = getOrCreateThreadPoolIfNecessary(datacenter, startThread);
-      if (replicaThreads == null) {
-        logger.warn("Number of replica threads is smaller or equal to 0, not starting any replica threads for {}.",
-            datacenter);
+      ReplicaThread replicaThread = getReplicaThreadForRemoteReplica(remoteReplicaInfo, startThread);
+      if (replicaThread == null) {
+        logger.warn(
+            "Number of replica threads is smaller or equal to 0 for data center {}, not starting any replica threads for {}.",
+            remoteReplicaInfo.getReplicaId().getDataNodeId().getDatacenterName(), remoteReplicaInfo.getReplicaId());
         continue;
       }
-      ReplicaThread replicaThread = dataNodeIdToReplicaThread.computeIfAbsent(dataNodeIdToReplicate,
-          key -> replicaThreads.get(getReplicaThreadIndexToUse(datacenter)));
       replicaThread.addRemoteReplicaInfo(remoteReplicaInfo);
       remoteReplicaInfo.setReplicaThread(replicaThread);
     }
   }
 
+  /**
+   * Get thread for given remote replica. Creates thread if it doesn't exist.
+   * @param remoteReplicaInfo {@link RemoteReplicaInfo} for which thread is being requested.
+   * @param startThread if thread needs to be started when created.
+   * @return {@link ReplicaThread} for this remote replica.
+   */
+  private ReplicaThread getReplicaThreadForRemoteReplica(RemoteReplicaInfo remoteReplicaInfo, boolean startThread) {
+    DataNodeId dataNodeIdToReplicate = remoteReplicaInfo.getReplicaId().getDataNodeId();
+    String datacenter = dataNodeIdToReplicate.getDatacenterName();
+    int numOfThreadsInPool =
+        datacenter.equals(dataNodeId.getDatacenterName()) ? replicationConfig.replicationNumOfIntraDCReplicaThreads
+            : replicationConfig.replicationNumOfInterDCReplicaThreads;
+    if (numOfThreadsInPool <= 0) {
+      return null;
+    }
+    if (!replicationConfig.replicationUseSeparateThreadForEachHost) {
```

Comment:
the order of conditions here will dictate what we prioritize - thread-per-host or the old-way